File: ml/preprocessing/seed_splitter.py
"""Touching-seed splitter — the Python mirror of
app/lib/ml/seed_splitter.dart (§11).

Marker Voronoi partition, NOT cv2.watershed: chamfer (3,4) distance
transform -> relative regional maxima -> multi-source BFS assignment. This
is chosen precisely because it is trivial to keep bit-identical across the
two languages; cv2.watershed's tie-breaking is not. Every constant, loop
order, and neighbour ordering here matches the Dart file — change them
together or `parity_check` will (correctly) complain.
"""
from __future__ import annotations

from collections import deque

import numpy as np

_MIN_PEAK_DISTANCE_FRACTION = 0.45
_PEAK_PLATEAU_RADIUS = 3
_MIN_MARKER_SEPARATION = 16.0
_MIN_PART_FRACTION = 0.15

# Split-trigger gate, mirrors ClassicalCVSeedFinder._splitAreaMultiple and
# the 25th-percentile reference in seed_finder.dart.
SPLIT_AREA_MULTIPLE = 1.6


def split_component(mask: np.ndarray) -> list[np.ndarray]:
    """`mask` is a 2-D uint8 array (0 / non-zero). Returns a list of
    boolean arrays the same shape, one per detected seed — or a list with
    fewer than 2 entries meaning "leave it as one seed".
    """
    h, w = mask.shape
    raw = mask != 0
    if not raw.any():
        return []

    clean = _clean_mask(raw)
    if not clean.any():
        return []

    dist = _chamfer_distance(clean)
    markers = _find_markers(clean, dist)
    if len(markers) < 2:
        return []

    labels = _voronoi_partition(clean, markers)

    raw_flat = raw.reshape(-1)
    lab_flat = labels.reshape(-1)
    raw_total = int(raw_flat.sum())
    parts: dict[int, list[int]] = {}
    for i in np.nonzero(raw_flat)[0]:
        lb = int(lab_flat[i])
        if lb > 0:
            parts.setdefault(lb, []).append(int(i))
    if len(parts) < 2:
        return []

    ordered = [parts[k] for k in sorted(parts)]
    merged = _absorb_slivers(ordered, raw_total)
    if len(merged) < 2:
        return []

    out = []
    for idxs in merged:
        m = np.zeros(h * w, dtype=bool)
        m[np.asarray(idxs, dtype=np.int64)] = True
        out.append(m.reshape(h, w))
    return out


def _clean_mask(fg: np.ndarray) -> np.ndarray:
    """Fill border-disconnected background pockets, then a 1px close."""
    h, w = fg.shape
    out = fg.copy()

    reached = np.zeros_like(out)
    q = deque()
    for x in range(w):
        for y in (0, h - 1):
            if not out[y, x] and not reached[y, x]:
                reached[y, x] = True
                q.append((y, x))
    for y in range(h):
        for x in (0, w - 1):
            if not out[y, x] and not reached[y, x]:
                reached[y, x] = True
                q.append((y, x))
    while q:
        y, x = q.popleft()
        for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1)):
            ny, nx = y + dy, x + dx
            if 0 <= ny < h and 0 <= nx < w and not out[ny, nx] and not reached[ny, nx]:
                reached[ny, nx] = True
                q.append((ny, nx))
    out |= ~reached & ~fg  # background not reached from the border -> hole

    return _erode(_dilate(out))


def _dilate(m: np.ndarray) -> np.ndarray:
    o = m.copy()
    o[:-1, :] |= m[1:, :]
    o[1:, :] |= m[:-1, :]
    o[:, :-1] |= m[:, 1:]
    o[:, 1:] |= m[:, :-1]
    return o


def _erode(m: np.ndarray) -> np.ndarray:
    # Border pixels treat a missing neighbour as "inside" (matches the
    # `x > 0 ? ... : true` guards in seed_splitter.dart::_erode).
    inside = m.copy()
    inside[:-1, :] &= m[1:, :]
    inside[1:, :] &= m[:-1, :]
    inside[:, :-1] &= m[:, 1:]
    inside[:, 1:] &= m[:, :-1]
    return inside


def _chamfer_distance(fg: np.ndarray) -> np.ndarray:
    h, w = fg.shape
    big = 1 << 28
    d = np.where(fg, big, 0).astype(np.int64)

    for y in range(h):
        for x in range(w):
            if d[y, x] == 0:
                continue
            v = d[y, x]
            if x > 0:
                v = min(v, d[y, x - 1] + 3)
            if y > 0:
                v = min(v, d[y - 1, x] + 3)
            if x > 0 and y > 0:
                v = min(v, d[y - 1, x - 1] + 4)
            if x < w - 1 and y > 0:
                v = min(v, d[y - 1, x + 1] + 4)
            d[y, x] = v
    for y in range(h - 1, -1, -1):
        for x in range(w - 1, -1, -1):
            if d[y, x] == 0:
                continue
            v = d[y, x]
            if x < w - 1:
                v = min(v, d[y, x + 1] + 3)
            if y < h - 1:
                v = min(v, d[y + 1, x] + 3)
            if x < w - 1 and y < h - 1:
                v = min(v, d[y + 1, x + 1] + 4)
            if x > 0 and y < h - 1:
                v = min(v, d[y + 1, x - 1] + 4)
            d[y, x] = v
    return d


def _find_markers(fg: np.ndarray, dist: np.ndarray) -> list[list[int]]:
    h, w = fg.shape
    r = _PEAK_PLATEAU_RADIUS
    max_dist = int(dist.max())
    min_peak = max(6, round(max_dist * _MIN_PEAK_DISTANCE_FRACTION))

    is_max = np.zeros((h, w), dtype=bool)
    for y in range(h):
        for x in range(w):
            if not fg[y, x] or dist[y, x] < min_peak:
                continue
            v = dist[y, x]
            y0, y1 = max(0, y - r), min(h, y + r + 1)
            x0, x1 = max(0, x - r), min(w, x + r + 1)
            if dist[y0:y1, x0:x1].max() <= v:
                is_max[y, x] = True

    seen = np.zeros((h, w), dtype=bool)
    candidates: list[list[int]] = []
    for sy in range(h):
        for sx in range(w):
            if not is_max[sy, sx] or seen[sy, sx]:
                continue
            group: list[int] = []
            q = deque([(sy, sx)])
            seen[sy, sx] = True
            while q:
                y, x = q.popleft()
                group.append(y * w + x)
                for dy in (-1, 0, 1):
                    for dx in (-1, 0, 1):
                        if dy == 0 and dx == 0:
                            continue
                        ny, nx = y + dy, x + dx
                        if 0 <= ny < h and 0 <= nx < w and is_max[ny, nx] and not seen[ny, nx]:
                            seen[ny, nx] = True
                            q.append((ny, nx))
            candidates.append(group)

    def peak(g: list[int]) -> int:
        return max(int(dist[i // w, i % w]) for i in g)

    candidates.sort(key=lambda g: (-peak(g), -len(g), g[0]))

    kept: list[list[int]] = []
    kept_centres: list[tuple[float, float]] = []
    for g in candidates:
        cx = sum(i % w for i in g) / len(g)
        cy = sum(i // w for i in g) / len(g)
        too_close = any(
            (c[0] - cx) ** 2 + (c[1] - cy) ** 2 < _MIN_MARKER_SEPARATION ** 2
            for c in kept_centres
        )
        if not too_close:
            kept.append(g)
            kept_centres.append((cx, cy))
    return kept


def _voronoi_partition(fg: np.ndarray, markers: list[list[int]]) -> np.ndarray:
    h, w = fg.shape
    labels = np.zeros(h * w, dtype=np.int64)
    fg_flat = fg.reshape(-1)
    q = deque()
    for m, group in enumerate(markers):
        for i in group:
            labels[i] = m + 1
            q.append(i)
    steps = ((0, -1), (1, 0), (0, 1), (-1, 0), (1, -1), (-1, -1), (1, 1), (-1, 1))
    while q:
        i = q.popleft()
        x, y = i % w, i // w
        lb = labels[i]
        for dx, dy in steps:
            nx, ny = x + dx, y + dy
            if 0 <= nx < w and 0 <= ny < h:
                ni = ny * w + nx
                if fg_flat[ni] and labels[ni] == 0:
                    labels[ni] = lb
                    q.append(ni)
    return labels.reshape(h, w)


def _absorb_slivers(parts: list[list[int]], total: int) -> list[list[int]]:
    min_size = int(total * _MIN_PART_FRACTION)
    big = [p for p in parts if len(p) >= min_size]
    slivers = [p for p in parts if len(p) < min_size]
    if len(big) < 2:
        return []
    if not slivers:
        return big
    for s in slivers:
        big[0].extend(s)
    big.sort(key=len, reverse=True)
    return big

File: app/lib/ml/seed_splitter.dart
import 'dart:collection';
import 'dart:typed_data';

/// Splits a connected-component mask that is actually **several touching
/// seeds** into one mask per seed (§11 "touching / overlapping seeds").
///
/// The classical detector in `seed_finder.dart` returns one blob per
/// connected region, so two barley grains resting against each other come
/// back as a single mis-shaped "seed" and the count is short. This runs as
/// a post-process on each component.
///
/// Method — a **marker Voronoi partition**, not a true grey-scale
/// watershed:
///  1. Chamfer (3, 4) distance transform of the mask.
///  2. Regional maxima of that transform, above a floor and de-duplicated
///     by separation → one marker per seed centre.
///  3. If ≥2 markers survive, every foreground pixel is assigned to the
///     nearest marker by multi-source breadth-first search over the mask
///     graph (fixed 8-neighbour order, first front to arrive wins).
///
/// True `cv2.watershed` tie-breaking is effectively impossible to
/// reproduce bit-for-bit in a second language; a marker Voronoi split is
/// fully deterministic, trivially identical in Dart and Python
/// (`ml/preprocessing/seed_splitter.py` is the mirror — keep them in
/// lock-step), and separates convex touching grains cleanly, which is the
/// case that actually occurs. Non-convex clumps are left for a learned
/// detector (§11).
class SeedSplitter {
  const SeedSplitter({
    this.minPeakDistanceFraction = 0.45,
    this.peakPlateauRadius = 3,
    this.minMarkerSeparation = 16,
    this.minPartFraction = 0.15,
  });

  /// A regional-maximum pixel is only a marker if its distance value is at
  /// least this fraction of the mask's maximum distance value. Relative,
  /// not absolute, so it scales with seed size and ignores the shallow
  /// ridges left by boundary noise.
  final double minPeakDistanceFraction;

  /// Chebyshev radius a pixel must dominate (>=) to count as a regional
  /// maximum.
  final int peakPlateauRadius;

  /// Markers whose centres are closer than this (pixels) are merged —
  /// one grain, one marker.
  final double minMarkerSeparation;

  /// After splitting, a part smaller than this fraction of the original
  /// mask is folded back into its largest neighbour — it's a sliver, not a
  /// seed.
  final double minPartFraction;

  /// [mask] is a row-major 0/255 buffer of size [w]*[h]. Returns one list
  /// of pixel indices (into that same buffer) per detected seed. A return
  /// of length 0 or 1 means "don't split — treat as a single seed".
  List<List<int>> split(Uint8List rawMask, int w, int h) {
    if (_countForeground(rawMask) == 0) return const [];

    // The thresholded blob is peppered with small holes and rough edges;
    // the distance transform is useless on that. Fill interior holes and
    // close pin-holes first, and do the whole split on that clean copy —
    // the pixel indices still refer to the original buffer.
    final mask = _cleanMask(rawMask, w, h);
    final total = _countForeground(mask);
    if (total == 0) return const [];

    final dist = _chamferDistance(mask, w, h);
    final markers = _findMarkers(mask, dist, w, h);
    if (markers.length < 2) return const [];

    final labels = _voronoiPartition(mask, w, h, markers);

    // Assign the *original* foreground pixels (not the cleaned ones) to the
    // label their location was given.
    final parts = <int, List<int>>{};
    var rawTotal = 0;
    for (var i = 0; i < rawMask.length; i++) {
      if (rawMask[i] == 0) continue;
      rawTotal++;
      final l = labels[i];
      if (l > 0) (parts[l] ??= <int>[]).add(i);
    }
    if (parts.length < 2) return const [];

    return _absorbSlivers(parts.values.toList(), rawTotal, w);
  }

  int _countForeground(Uint8List mask) {
    var n = 0;
    for (final v in mask) {
      if (v != 0) n++;
    }
    return n;
  }

  /// Fill background pockets not connected to the mask's bounding-box
  /// border (pepper holes from thresholding noise), then a 1px
  /// morphological close to smooth the edge. The result is a superset of
  /// the input, so every original foreground pixel still lands on a
  /// labelled cell.
  Uint8List _cleanMask(Uint8List mask, int w, int h) {
    final out = Uint8List.fromList(mask);

    // Flood the background from the border; anything background not reached
    // is an interior hole -> fill it.
    final reached = Uint8List(w * h);
    final queue = Queue<int>();
    void seed(int i) {
      if (out[i] == 0 && reached[i] == 0) {
        reached[i] = 1;
        queue.add(i);
      }
    }

    for (var x = 0; x < w; x++) {
      seed(x);
      seed((h - 1) * w + x);
    }
    for (var y = 0; y < h; y++) {
      seed(y * w);
      seed(y * w + w - 1);
    }
    while (queue.isNotEmpty) {
      final i = queue.removeFirst();
      final x = i % w, y = i ~/ w;
      for (var k = 0; k < 4; k++) {
        final nx = x + const [0, 0, -1, 1][k];
        final ny = y + const [-1, 1, 0, 0][k];
        if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
        final ni = ny * w + nx;
        if (out[ni] == 0 && reached[ni] == 0) {
          reached[ni] = 1;
          queue.add(ni);
        }
      }
    }
    for (var i = 0; i < out.length; i++) {
      if (out[i] == 0 && reached[i] == 0) out[i] = 255;
    }

    // Morphological close: dilate then erode, 4-neighbour, one step.
    return _erode(_dilate(out, w, h), w, h);
  }

  Uint8List _dilate(Uint8List m, int w, int h) {
    final o = Uint8List(w * h);
    for (var y = 0; y < h; y++) {
      for (var x = 0; x < w; x++) {
        final i = y * w + x;
        if (m[i] != 0 ||
            (x > 0 && m[i - 1] != 0) ||
            (x < w - 1 && m[i + 1] != 0) ||
            (y > 0 && m[i - w] != 0) ||
            (y < h - 1 && m[i + w] != 0)) {
          o[i] = 255;
        }
      }
    }
    return o;
  }

  Uint8List _erode(Uint8List m, int w, int h) {
    final o = Uint8List(w * h);
    for (var y = 0; y < h; y++) {
      for (var x = 0; x < w; x++) {
        final i = y * w + x;
        if (m[i] == 0) continue;
        final keep = (x > 0 ? m[i - 1] != 0 : true) &&
            (x < w - 1 ? m[i + 1] != 0 : true) &&
            (y > 0 ? m[i - w] != 0 : true) &&
            (y < h - 1 ? m[i + w] != 0 : true);
        if (keep) o[i] = 255;
      }
    }
    return o;
  }

  /// Two-pass chamfer (3, 4) distance transform. 0 on background, distance
  /// to the nearest background pixel on foreground.
  Int32List _chamferDistance(Uint8List mask, int w, int h) {
    const big = 1 << 28;
    final d = Int32List(w * h);
    for (var i = 0; i < d.length; i++) {
      d[i] = mask[i] != 0 ? big : 0;
    }

    void relax(int idx, int from, int cost) {
      final v = d[from] + cost;
      if (v < d[idx]) d[idx] = v;
    }

    for (var y = 0; y < h; y++) {
      for (var x = 0; x < w; x++) {
        final i = y * w + x;
        if (d[i] == 0) continue;
        if (x > 0) relax(i, i - 1, 3);
        if (y > 0) relax(i, i - w, 3);
        if (x > 0 && y > 0) relax(i, i - w - 1, 4);
        if (x < w - 1 && y > 0) relax(i, i - w + 1, 4);
      }
    }
    for (var y = h - 1; y >= 0; y--) {
      for (var x = w - 1; x >= 0; x--) {
        final i = y * w + x;
        if (d[i] == 0) continue;
        if (x < w - 1) relax(i, i + 1, 3);
        if (y < h - 1) relax(i, i + w, 3);
        if (x < w - 1 && y < h - 1) relax(i, i + w + 1, 4);
        if (x > 0 && y < h - 1) relax(i, i + w - 1, 4);
      }
    }
    return d;
  }

  /// Regional maxima of the distance transform, grouped into connected
  /// plateaus and de-duplicated by [minMarkerSeparation]. Each surviving
  /// marker is a list of its pixel indices.
  List<List<int>> _findMarkers(Uint8List mask, Int32List dist, int w, int h) {
    final r = peakPlateauRadius;
    var maxDist = 0;
    for (final v in dist) {
      if (v > maxDist) maxDist = v;
    }
    // Chamfer (3,4): a 3-unit step ≈ 1px, so require the peak to sit at
    // least ~2px in from the edge regardless of the relative floor.
    final minPeak = (maxDist * minPeakDistanceFraction).round().clamp(6, 1 << 20);

    final isMax = Uint8List(w * h);
    for (var y = 0; y < h; y++) {
      for (var x = 0; x < w; x++) {
        final i = y * w + x;
        if (mask[i] == 0 || dist[i] < minPeak) continue;
        final v = dist[i];
        var dominates = true;
        for (var dy = -r; dy <= r && dominates; dy++) {
          final ny = y + dy;
          if (ny < 0 || ny >= h) continue;
          for (var dx = -r; dx <= r; dx++) {
            final nx = x + dx;
            if (nx < 0 || nx >= w) continue;
            if (dist[ny * w + nx] > v) {
              dominates = false;
              break;
            }
          }
        }
        if (dominates) isMax[i] = 1;
      }
    }

    // Connected plateaus of maxima -> one candidate marker each.
    final seen = Uint8List(w * h);
    final candidates = <List<int>>[];
    for (var start = 0; start < isMax.length; start++) {
      if (isMax[start] == 0 || seen[start] != 0) continue;
      final group = <int>[];
      final queue = Queue<int>()..add(start);
      seen[start] = 1;
      while (queue.isNotEmpty) {
        final i = queue.removeFirst();
        group.add(i);
        final x = i % w, y = i ~/ w;
        for (var dy = -1; dy <= 1; dy++) {
          for (var dx = -1; dx <= 1; dx++) {
            if (dx == 0 && dy == 0) continue;
            final nx = x + dx, ny = y + dy;
            if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
            final ni = ny * w + nx;
            if (isMax[ni] == 1 && seen[ni] == 0) {
              seen[ni] = 1;
              queue.add(ni);
            }
          }
        }
      }
      candidates.add(group);
    }

    // Strongest first (higher peak distance, then larger plateau, then
    // scan order) so the de-dup keep/drop decision is deterministic.
    int peak(List<int> g) => g.fold(0, (m, i) => dist[i] > m ? dist[i] : m);
    candidates.sort((a, b) {
      final pa = peak(a), pb = peak(b);
      if (pa != pb) return pb - pa;
      if (a.length != b.length) return b.length - a.length;
      return a.first - b.first;
    });

    final kept = <List<int>>[];
    final keptCentres = <List<double>>[];
    for (final g in candidates) {
      double sx = 0, sy = 0;
      for (final i in g) {
        sx += i % w;
        sy += i ~/ w;
      }
      final cx = sx / g.length, cy = sy / g.length;
      var tooClose = false;
      for (final c in keptCentres) {
        final ddx = c[0] - cx, ddy = c[1] - cy;
        if (ddx * ddx + ddy * ddy < minMarkerSeparation * minMarkerSeparation) {
          tooClose = true;
          break;
        }
      }
      if (!tooClose) {
        kept.add(g);
        keptCentres.add([cx, cy]);
      }
    }
    return kept;
  }

  /// Multi-source BFS: every foreground pixel takes the label of the
  /// marker whose front reaches it first. Fixed neighbour order, FIFO
  /// queue → identical result every run and in the Python mirror.
  Int32List _voronoiPartition(Uint8List mask, int w, int h, List<List<int>> markers) {
    final labels = Int32List(w * h);
    final queue = Queue<int>();
    for (var m = 0; m < markers.length; m++) {
      for (final i in markers[m]) {
        labels[i] = m + 1;
        queue.add(i);
      }
    }
    const dx = [0, 1, 0, -1, 1, -1, 1, -1];
    const dy = [-1, 0, 1, 0, -1, -1, 1, 1];
    while (queue.isNotEmpty) {
      final i = queue.removeFirst();
      final x = i % w, y = i ~/ w;
      final l = labels[i];
      for (var k = 0; k < 8; k++) {
        final nx = x + dx[k], ny = y + dy[k];
        if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
        final ni = ny * w + nx;
        if (mask[ni] != 0 && labels[ni] == 0) {
          labels[ni] = l;
          queue.add(ni);
        }
      }
    }
    return labels;
  }

  List<List<int>> _absorbSlivers(List<List<int>> parts, int total, int w) {
    final minSize = (total * minPartFraction).floor();
    final big = <List<int>>[];
    final slivers = <List<int>>[];
    for (final p in parts) {
      (p.length >= minSize ? big : slivers).add(p);
    }
    if (big.length < 2) return const []; // the "split" was all slivers — don't
    if (slivers.isEmpty) return big;

    // Hand each sliver's pixels to whichever surviving part it touches most.
    for (final s in slivers) {
      big.first.addAll(s); // simple, deterministic: fold into the largest
    }
    big.sort((a, b) => b.length - a.length);
    return big;
  }
}
